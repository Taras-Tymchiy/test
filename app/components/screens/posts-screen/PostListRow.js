// @flow
import * as React from 'react';
import { View } from 'react-native';
import { Post } from '../../../entities/Post';
import FormattedDate from '../../atoms/FormattedDate';
import ListRow from '../../atoms/ListRow';
import { TextPrimary, TextSecondary, TextRegular } from '../../atoms/Text';
import styles from './PostListRow.styles';

interface PostListRowProps {
    post: Post;
}

export default function PostListRow({post}: PostListRowProps) {
  const { id, created_at, user, text } = post;
  return (
    <ListRow>
      <View style={styles.header}> 
        <TextPrimary>{user.name}</TextPrimary>
        <TextSecondary>
          <FormattedDate value={created_at} />
        </TextSecondary>
      </View>
      <TextRegular>{text}</TextRegular>
    </ListRow>
  );
};
