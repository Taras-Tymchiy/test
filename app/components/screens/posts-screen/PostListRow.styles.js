// @flow
import { StyleSheet } from 'react-native';
import { SPACING_S } from '../../styles/constants';

export default StyleSheet.create({
  header: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: SPACING_S
  }
});