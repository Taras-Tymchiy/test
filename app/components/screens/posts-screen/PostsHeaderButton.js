// @flow
import * as React from 'react';
import SimpleButton from '../../atoms/SimpleButton';
import StyledHOC from '../../HOCs/StyledHOC';
import { SPACING_M } from '../../styles/constants';

export default StyledHOC(SimpleButton, { marginRight: SPACING_M })
