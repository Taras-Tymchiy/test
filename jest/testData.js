import { Post } from "../app/entities/Post";


export const posts: Post[] = [
    {
        id: 1,
        id_str: '1',
        created_at: '2015-01-01',
        user: {
            id: 1,
            name: 'John Snow',
        },
        text: 'post text 1'
    },
    {
        id: 2,
        id_str: '2',
        created_at: '2015-01-02',
        user: {
            id: 2,
            name: 'Barack Obama',
        },
        text: 'post text 2'
    }
]