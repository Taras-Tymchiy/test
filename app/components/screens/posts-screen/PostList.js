// @flow
import * as React from 'react';
import { FlatList } from 'react-native';
import { Post } from '../../../entities/Post';
import PostListRow from './PostListRow';

interface PostListProps {
  posts: Post[];
}

const keyExtractor = (post: Post, index) => post.id_str;

export default ({ posts }: PostListProps) => 
  <FlatList
    data={posts}
    keyExtractor={keyExtractor}
    renderItem={({item}) => <PostListRow post={item} />}
  />;
