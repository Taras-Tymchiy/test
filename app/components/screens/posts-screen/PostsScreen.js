// @flow
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect, Dispatch } from 'react-redux';
import * as postsActions from '../../../store/actions/PostActions';
import { type PostsState } from '../../../store/reducers/postsReducer';
import { Post } from '../../../entities/Post';
import PostList from './PostList';
import PostsHeaderButton from './PostsHeaderButton';
import { type RootState } from '../../../store/reducers/RootState';
import SpinnerHOC from '../../HOCs/SpinnerHOC';
import ErrorHOC from '../../HOCs/ErrorHOC';

export interface PostsScreenProps {
  +posts: Post[];
  +isLoading: boolean;
  +isPulling: boolean;
  +error: Error;
  +pullInterval: number;
  +postsCount: number;
  +postsUrl: string;
  +isCurrentView: boolean;
  +actions: typeof postsActions;
}

const PostsListsEnhanced = SpinnerHOC(ErrorHOC(PostList));

// exporting for testing purposes
export class PostsScreen extends Component<PostsScreenProps> {

  static navigationOptions = ({ navigation }) => ({
    headerRight: <PostsHeaderButton onPress={()=> navigation.navigate('Settings')} title="Settings" />,
  });

  render() {
    const { posts, isLoading, error } = this.props;
    return <PostsListsEnhanced posts={posts} spinner={isLoading && !posts.length} error={error} />
  }

  startPulling(props: PostsScreenProps) {
    const { actions, pullInterval, postsUrl, isPulling, postsCount } = props;
    if (!isPulling) {
      actions.startPullingPosts({url: postsUrl, count: postsCount}, pullInterval);
    }
  }
  
  stopPulling() {
    this.props.actions.stopPullingPosts();
  }

  componentWillUnmount() {
    this.stopPulling();
  }
  
  componentDidMount() {
    this.startPulling(this.props);
  }

  componentWillReceiveProps(nextProps: PostsScreenProps) {
    // navigation events will be implemented in react-navigation very soon
    // see the PR https://github.com/react-navigation/react-navigation/pull/3345
    // but for now will use some dirty tricks to determine whether current page is active
    if (nextProps.isCurrentView) {
      this.startPulling(nextProps);
    } else {
      this.stopPulling();
    }
  }

}

function mapDispatchToProps(dispatch: Dispatch<{}>) {
  return {
      actions: bindActionCreators(postsActions, dispatch)
  };
}

function mapStateToProps(state: RootState) {
  return {
      posts: state.posts.data,
      isPulling: state.posts.isPulling,
      isLoading: state.posts.isLoading,
      error: state.posts.error,
      postsUrl: state.settings.postsUrl,
      postsCount: state.settings.postsCount,
      pullInterval: state.settings.pullInterval,
      isCurrentView: state.navigation.index === 0 
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(PostsScreen);
