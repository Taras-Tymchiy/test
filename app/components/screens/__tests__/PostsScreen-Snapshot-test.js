import React from 'react';
import renderer from 'react-test-renderer';
import { PostsScreen } from '../posts-screen/PostsScreen';
import { posts } from '../../../../jest/testData';

jest.mock('../posts-screen/PostList', (props) => `PostList ${props}`);

const props = {
  posts,
  isLoading: false,
  syncInProgess: false,
  error: null,
  pullIntervl: 5000,
  postsCount: 10,
  postsUrl: 'test',
  isCurrentView: false,
  actions: {
    startSync: () => null,
    stopSync: () => null
  }
};

it('renders posts in a PostsScreen using Snapshots', () => {
  expect(renderer.create(
    <PostsScreen {...props} />
  )).toMatchSnapshot();
});
