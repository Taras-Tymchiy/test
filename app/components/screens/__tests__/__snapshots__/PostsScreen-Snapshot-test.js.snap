// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders posts in a PostsScreen using Snapshots 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <PostList undefined
      posts={
        Array [
          Object {
            "created_at": "2015-01-01",
            "id": 1,
            "id_str": "1",
            "text": "post text 1",
            "user": Object {
              "id": 1,
              "name": "John Snow",
            },
          },
          Object {
            "created_at": "2015-01-02",
            "id": 2,
            "id_str": "2",
            "text": "post text 2",
            "user": Object {
              "id": 2,
              "name": "Barack Obama",
            },
          },
        ]
      }
    />
  </View>
</View>
`;
